plugins/processors/aws_ec2/ec2.go

```diff
@@ -208,31 +207,99 @@ func (r *AwsEc2Processor) LookupIMDSTags(metric telegraf.Metric) telegraf.Metric
 		return metric
 	}
 
-	iido, err := r.imdsClient.GetInstanceIdentityDocument(
-		ctx,
-		&imds.GetInstanceIdentityDocumentInput{},
-	)
+	doc, err := r.imdsClient.GetInstanceIdentityDocument(ctx, &imds.GetInstanceIdentityDocumentInput{})
+	if err != nil {
+		r.Log.Errorf("Error when calling GetInstanceIdentityDocument: %v", err)
+		return metric
+	}
+
+	for _, tag := range tagsNotFound {
+		var v string
+		switch tag {
+		case "accountId":
+			v = doc.AccountID
+		case "architecture":
+			v = doc.Architecture
+		case "availabilityZone":
+			v = doc.AvailabilityZone
+		case "billingProducts":
+			v = strings.Join(doc.BillingProducts, ",")
+		case "imageId":
+			v = doc.ImageID
+		case "instanceId":
+			v = doc.InstanceID
+		case "instanceType":
+			v = doc.InstanceType
+		case "kernelId":
+			v = doc.KernelID
+		case "pendingTime":
+			v = doc.PendingTime.String()
+		case "privateIp":
+			v = doc.PrivateIP
+		case "ramdiskId":
+			v = doc.RamdiskID
+		case "region":
+			v = doc.Region
+		case "version":
+			v = doc.Version
+		default:
+			continue
+		}
+
+		metric.AddTag(tag, v)
+		expiration := int(time.Duration(r.CacheTTL).Seconds())
+		if err := r.tagCache.Set([]byte(tag), []byte(v), expiration); err != nil {
+			r.Log.Errorf("Error when setting IMDS tag cache value: %v", err)
+			continue
+		}
+	}
+
+	return metric
+}
+
+func (r *AwsEc2Processor) lookupMetadataTags(metric telegraf.Metric) telegraf.Metric {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(r.Timeout))
+	defer cancel()
+
+	var tagsNotFound []string
+	for _, tag := range r.MetadataTags {
+		val, err := r.tagCache.Get([]byte(tag))
+		if err != nil {
+			tagsNotFound = append(tagsNotFound, tag)
+		} else {
+			metric.AddTag(tag, string(val))
+		}
+	}
 
+	if len(tagsNotFound) == 0 {
+		return metric
+	}
+
+	metadata, err := r.imdsClient.GetMetadata(ctx, &imds.GetMetadataInput{})
```

Comment:
I think it can be hardcoded with `tags/instance`

Reply:
Path can be empty to retrieve all as I read the documentation...

Reply:
As documentation says:
> 	// The relative metadata path to retrieve. Can be empty string to retrieve
	// a response containing a new line separated list of metadata resources
	// available.

So I read this as: It's a potential filter and then you get _every_ metadata resource, won't you?

Reply:
Yes, you're right, but do we need to retrieve every metadata resource? Maybe it's better to filter them

Anyway, I don't know why it didn't work for me :( 

https://github.com/influxdata/telegraf/issues/15789#issuecomment-2320329494

Reply:
I added a trace message, could you please add `log_level = "trace"` to the processor!? It should print the received metadata output...

Reply:
Seems like you are right, we have to specify a path. Please test the modified binary once available...

Reply:
@srebhan one test has been failed, could you please take a look?

Reply:
Hmmm very strange. I would have expected that the call provides _all_ tags for the instance. Could you post the (redacted) output of a curl call to the `tags/instance` path to see what we get? Otherwise I can try with the _full_ path to the actual tag but this will cause one GET request per tag I fear...

Reply:
@srebhan 

```
# curl -H "X-aws-ec2-metadata-token: $TOKEN" ${AWS_API_URL}/latest/meta-data/tags/instance/
Name
Project
Stage
```

Reply:
@srebhan any news? :)

Reply:
Sorry for not finishing this yesterday. Please find an updated version below. I found out that the response is in the `Content` field and is not parsed yet... Anyway, I also updated the configuration to accept full metadata paths as I can see people wanting to use e.g. the `hostname` or IP for tagging. So please now use the new `metadata_paths` option with `metadata_paths = ["tags/instance/Name", "tags/instance/Stage"]`...

Reply:
@srebhanI can confirm that it works for me now, thank you for your work and I think it was a good idea to use full metadata paths

do you know when it can be available with new version of telegraf?

Reply:
If you are lucky we might be able to include it in v1.32.0 which is due on Monday 2024-09-09. Otherwise you might need to wait for v1.33.0...